posts/2021-09-03-microprofile-21009.adoc

```diff
@@ -0,0 +1,268 @@
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/18009[Wrong char count in ServletOutputStream with non-ASCII characters skips content]
++
+A bug was discovered where the `ServletOutputStream` would skip characters when using `print` or `println` to print content with multi-byte non-ASCII characters. This bug was fixed by changing the `WCOutputStream` to write the length of the encoded bytes rather than the length of the original string.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/18002[@Schema(multipleOf = ) throws NumberFormatException in mpOpenAPI-2.0 feature]
++
+In the `mpOpenAPI-2.0` feature, a `NumberFormatException` could be incorrectly thrown when using the `multipleOf` attribute of the `@org.eclipse.microprofile.openapi.annotations.media.Schema` annotation. With `multipleOf = 1`, the server shows:
+
+[source, java]
+----
+[INFO] [ERROR ] CWWKO1650E: Validation of the OpenAPI document produced the following error(s):
+[INFO]
+[INFO] - Message: The Schema Object must have the "multipleOf" property set to a number strictly greater than zero, Location: #/components/schemas/...
+----
+
+This issue was fixed by updating the `multipleOf` validation check, to ensure that it is checking that `multipleOf` is greater than zero, rather than one.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17682[Exception stack trace is exposed in error returns from JMX REST APIs]
++
+When Liberty's `JMX` REST APIs encounter errors, a JSON object is returned which has a string field labelled 'stackTrace' containing the stack trace from the exception which caused the error. Returning the stack trace was flagged as a security concern, so this field has been removed. An extra string field labelled `error` has been added to the returned `JSON` object, containing the message from the java exception.
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17912[New Netty 4.1.66 release]
++
+We've pulled in the latest release of the Netty project, which is a dependency of our grpcClient-1.0 feature. This new release includes multiple improvements and bugfixes, for more information see the link:https://netty.io/news/2021/07/16/4-1-66-Final.html[Netty release notes].
+
+* link:https://github.com/OpenLiberty/open-liberty/issues/17444[Incorrect Expression Language (EL) Method Matching with Varargs]
++
+An `el-3.0` performance patch, for Open Liberty issue link:https://github.com/OpenLiberty/open-liberty/issues/14175[#14175], in 21.0.0.3 inadvertently ported over new `varargs` matching code, which later exposed a bug in that code. The EL implementation would select the wrong method if `varargs` were used when overloading a method.  The code has since been updated upstream in the Tomcat community (see Bugzilla Issue 65358), and it has also been port over to Liberty. 
```

Comment:
An `el-3.0` performance patch, for Open Liberty issue link:https://github.com/OpenLiberty/open-liberty/issues/14175[#14175], **released in** 21.0.0.3 inadvertently ported over new `varargs` matching code, which later exposed a bug in that code. The EL implementation **could** select the wrong method if `varargs` were used when overloading a method. The code has since been updated upstream in the Tomcat community (see Bugzilla Issue 65358), and it has also been **ported** over to Liberty.